File: apps/web/src/types/tactic.ts
export type Duty = "Attack" | "Support" | "Defend" | "Automatic" | "Cover" | "Stopper";
export type FmVersion = "FM23" | "FM24";

export interface PlayerPosition {
  id: string;
  role: string;           // e.g., "Striker", "Deep Lying Playmaker"
  duty: Duty;
  x: number;              // 0–100 (% of pitch width, left to right)
  y: number;              // 0–100 (% of pitch height, own goal = 100)
  jerseyNumber?: number;
  name?: string;
}

export interface TacticArrow {
  id: string;
  fromPlayerId: string;
  toPlayerId: string;
  type: "movement" | "press" | "support" | "overlap";
}

export interface Tactic {
  id?: string;
  title: string;
  formation: string;
  style: string;
  mentality: "Defensive" | "Cautious" | "Balanced" | "Positive" | "Attacking";
  inPossession: Record<string, string | boolean>;
  inTransition: Record<string, string | boolean>;
  outOfPossession: Record<string, string | boolean>;
  players: PlayerPosition[];
  arrows: TacticArrow[];
  fmVersion?: FmVersion;
  isPublic?: boolean;
}

export interface ChannelOccupation {
  wideLeft: number;
  halfSpaceLeft: number;
  center: number;
  halfSpaceRight: number;
  wideRight: number;
}

/** A pairwise chemistry relationship between two players */
export interface Synergy {
  player1Id: string;
  player2Id: string;
  /** positive = green bond, negative = clash, tension = amber warning */
  type: "positive" | "negative" | "tension";
  /** 0–100 chemistry score */
  score: number;
  /** Short name, e.g. "Elite Pivot" */
  label: string;
  /** Full coaching description for tooltip */
  description: string;
}

/** A named three-player tactical unit */
export interface CompatibilityTriangle {
  player1Id: string;
  player2Id: string;
  player3Id: string;
  /** Composite score = average of three pair scores */
  score: number;
  /** Pattern name, e.g. "Wide Overlap Triangle" */
  label: string;
  description: string;
}

export interface RiskFactor {
  area: "left_flank" | "right_flank" | "central" | "defence" | "attack";
  severity: "critical" | "warning";
  message: string;
}

export interface DutyBalance {
  defend: number;
  support: number;
  attack: number;
}

export interface AnalysisResult {
  tacticalNarrative: string;
  inPossessionRating: number;
  outOfPossessionRating: number;
  channelOccupation: ChannelOccupation;
  restDefenceStructure: string;
  buildUpStructure: string;
  dutyBalance: DutyBalance;
  penetration: number;
  solidity: number;
  /** Pairwise role chemistry bonds */
  synergies: Synergy[];
  /** Named three-player tactical units */
  compatibilityTriangles: CompatibilityTriangle[];
  riskFactors: RiskFactor[];
  suggestions: Suggestion[];
}

export interface Suggestion {
  severity: "critical" | "warning" | "positive";
  area: "left_flank" | "right_flank" | "central" | "defence" | "attack" | "team";
  message: string;
}

File: packages/tacticlab-core/src/models/tactic.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct PlayerPosition {
    pub id: String,
    pub role: String,
    pub duty: String,
    pub x: f32,
    pub y: f32,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct TacticArrow {
    pub id: String,
    pub from_player_id: String,
    pub to_player_id: String,
    pub arrow_type: String,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct Tactic {
    pub title: String,
    pub formation: String,
    pub style: String,
    pub mentality: String,
    pub in_possession: HashMap<String, serde_json::Value>,
    pub in_transition: HashMap<String, serde_json::Value>,
    pub out_of_possession: HashMap<String, serde_json::Value>,
    pub players: Vec<PlayerPosition>,
    pub arrows: Vec<TacticArrow>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct ChannelOccupation {
    pub wide_left: f32,
    pub half_space_left: f32,
    pub center: f32,
    pub half_space_right: f32,
    pub wide_right: f32,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct Synergy {
    pub player1_id: String,
    pub player2_id: String,
    pub r#type: String,   // "positive" | "negative" | "tension"
    pub score: u32,       // 0–100
    pub label: String,
    pub description: String,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct CompatibilityTriangle {
    pub player1_id: String,
    pub player2_id: String,
    pub player3_id: String,
    pub score: u32,
    pub label: String,
    pub description: String,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct RiskFactor {
    pub area: String,
    pub severity: String,
    pub message: String,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct Suggestion {
    pub severity: String,
    pub area: String,
    pub message: String,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct DutyBalance {
    pub defend: u32,
    pub support: u32,
    pub attack: u32,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct AnalysisResult {
    pub tactical_narrative: String,
    pub in_possession_rating: f32,
    pub out_of_possession_rating: f32,
    pub channel_occupation: ChannelOccupation,
    pub rest_defence_structure: String,
    pub build_up_structure: String,
    pub duty_balance: DutyBalance,
    pub penetration: f32,
    pub solidity: f32,
    pub synergies: Vec<Synergy>,
    pub compatibility_triangles: Vec<CompatibilityTriangle>,
    pub risk_factors: Vec<RiskFactor>,
    pub suggestions: Vec<Suggestion>,
}
